controller/orphan_controller.go

```diff
@@ -0,0 +1,387 @@
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync orphan %v", oc.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != oc.namespace {
+		return nil
+	}
+	return oc.reconcile(name)
+}
+
+func (oc *OrphanController) reconcile(orphanName string) (err error) {
+	orphan, err := oc.ds.GetOrphan(orphanName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	log := getLoggerForOrphan(oc.logger, orphan)
+
+	if !oc.isResponsibleFor(orphan) {
+		return nil
+	}
+
+	if orphan.Status.OwnerID != oc.controllerID {
+		orphan.Status.OwnerID = oc.controllerID
+		orphan, err = oc.ds.UpdateOrphanStatus(orphan)
+		if err != nil {
+			// we don't mind others coming first
+			if apierrors.IsConflict(errors.Cause(err)) {
+				return nil
+			}
+			return err
+		}
+		log.Infof("Orphan Controller %v picked up %v", oc.controllerID, orphan.Name)
+	}
+
+	if !orphan.DeletionTimestamp.IsZero() {
+		defer func() {
+			if err == nil {
+				err = oc.ds.RemoveFinalizerForOrphan(orphan)
+			}
+		}()
+
+		return oc.deleteOrphanedData(orphan)
+	}
+
+	existingOrphan := orphan.DeepCopy()
+	defer func() {
+		if err != nil {
+			return
+		}
+		if reflect.DeepEqual(existingOrphan.Status, orphan.Status) {
+			return
+		}
+		if _, err := oc.ds.UpdateOrphanStatus(orphan); err != nil && apierrors.IsConflict(errors.Cause(err)) {
+			log.WithError(err).Debugf("Requeue %v due to conflict", orphanName)
+			oc.enqueueOrphan(orphan)
+		}
+	}()
+
+	if err := oc.updateConditions(orphan); err != nil {
+		log.WithError(err).Errorf("failed to update conditions for orphan %v", orphan.Name)
+		return err
+	}
+
+	return nil
+}
+
+func getLoggerForOrphan(logger logrus.FieldLogger, orphan *longhorn.Orphan) *logrus.Entry {
+	return logger.WithFields(
+		logrus.Fields{
+			"orphan": orphan.Name,
+		},
+	)
+}
+
+func (oc *OrphanController) isResponsibleFor(orphan *longhorn.Orphan) bool {
+	return isControllerResponsibleFor(oc.controllerID, oc.ds, orphan.Name, orphan.Spec.NodeID, orphan.Status.OwnerID)
+}
+
+func (oc *OrphanController) deleteOrphanedData(orphan *longhorn.Orphan) (err error) {
+	log := getLoggerForOrphan(oc.logger, orphan)
+
+	defer func() {
+		if err == nil {
+			return
+		}
+
+		log.WithError(err).Errorf("error deleting orphan %v data", orphan.Name)
+		orphan.Status.Conditions = types.SetCondition(orphan.Status.Conditions,
+			longhorn.OrphanConditionTypeError, longhorn.ConditionStatusTrue, "", err.Error())
+	}()
+
+	// Make sure if the orphan nodeID and controller ID are same.
+	// If NO, just delete the orphan resource object and don't touch the data.
+	if orphan.Spec.NodeID != oc.controllerID {
+		log.Infof("Orphan nodeID %v is different from controllerID %v, so just delete the orphan resouce object",
+			orphan.Name, oc.controllerID)
+		return nil
+	}
+
+	if types.GetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeDataCleanable).Status ==
```

Comment:
Question: what if the Orphan CR is being deleted while the condition field is still empty?

Reply:
Good catch!
For CR operations, can the deletion happen before the creation?
If YES, I incline to delete the orphan CR only and don't touch the orphaned data.

Reply:
So, the following change will be better.

```
	if types.GetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeDataCleanable).Status !=
		longhorn.ConditionStatusTrue {
		log.Infof("Only delete orphan %v resource object and do not delete the orphaned data", orphan.Name)
		return nil
	}
```

Reply:
The CR can be marked as deletion immediately after the creation. For example, when the auto-deletion setting is enabled, will the node controller wait for the condition update before sending the deletion request?

Reply:
No, the node controller won't wait for it. You are right. The node controller should check the condition before sending the deletion.